design/sync-labels-bmh-to-node.md

```diff
@@ -0,0 +1,217 @@
+
+We need to expose two additional configuration variables to the user: the set of label prefixes and the full-sync interval. We can introduce two new command-line flags for CAPM3, and also utilize default values if the flags are not specified: (i) `--bmh-label-prefixes` with a nil default value and (ii) `--label-sync-interval` with a 15-second default value. 
+
+Unfortunately, command-line flags will not be compatible with openShift where cluster administrators do not deploy metal3 themselves and do not have access to the command line arguments for the cluster-api provider. 
+
+### Annotations on BMH
+
+In this approach, an annotation is added to a BMH, in addition to the labels to be synchronized, which reflects the prefixes to be utilized. For example,
+
+```
+  kind: BareMetalHost
+  metadata:
+    annotations:   
+      cluster.x-k8s.io/label-prefixes: "my-prefix.metal3.io, my-prefix.kubernetes.io"
+      cluster.x-k8s.io/label-sync-interval: "30s"   
+    labels:
+      my-prefix.metal3.io/rack: xyz-123
+      my-prefix.kubernetes.io/zone: security-level-0   
+```
+
+A concern with this approach is consistency since the same annotation would have to be duplicated across BMHs. The prefixes, while user configurable, are static and should not require re-configuring. For example, in a likely scenario, the cluster admin would define a standard set of prefixes to be used internally. Also, annotations can be (accidentally) added/removed/changed, leading to undesirable behavior. 
+
+### New API Type
+
+We could introduce a new API type: `BareMetalHostLabelSyncProfile` (any other suggestions for the name?).
+
+The common case is when a label on the BMH is updated, we must first find the corresponding `BareMetalHostLabelSyncProfile` (if any) that matches the BMH. To avoid fetching all `BareMetalHostLabelSyncProfile`, we can make use of a label on the BMH, in namespace/name format, that references the `BareMetalHostLabelSyncProfile` (see example below). We can drop the namespace requirement if it's in the same namespace as the BMH.
+
+The less common case is where the `BareMetalHostLabelSyncProfile` is updated (e.g. a prefix is removed). In this case, we will need to fetch all BMHs associated and perform a sync operations across the impacted Nodes. Label selectors could be used for this purpose. 
+
+The approach is illustrated below:
+
+```
+kind: BareMetalHost
+metadata:
+  labels:
+    metal3.io/baremetalhost-label-sync-profile: some-ns/label-sync-profile-0
+...
+---
+kind: BareMetalHostLabelSyncProfile
+name: label-sync-profile-0
+namespace: some-ns
+labelSelector: 
+  matchLabels: 
+    metal3.io/baremetalhost-label-sync-profile: some-ns/label-sync-profile-0
+prefixes:
+  - my-prefix.metal3.io
+  - my-prefix.kubernetes.io
+label-sync-interval: "30s"  
+```
+
+It's the admin's/user's responsibility to create the linkage by labeling BMHs (e.g `metal3.io/baremetalhost-label-sync-profile: some-ns/label-sync-profile-0`). 
+
+### Metal3Cluster 
```

Comment:
@maelk I'm also inclined to use Metal3Cluster as the top level object where prefixes get specified. Based on your feedback, I have updated the proposal with how the BMH <--> Metal3Cluster mapping may work. PTAL.

@dhellmann What to do you think about using Metal3Cluster instead of a new API type?